commit 72a8fd401f4c122932be0d854761a934fd3b7f25
Author: Michael Still <mikal@stillhq.com>
Date:   Sun Jul 21 10:29:24 2024 +1000

    This is WIP and doesn't work yet. Hence not being included in the patch
    order.

diff --git a/nova/db/main/migrations/versions/2903cd72dc14_add_tls_port_to_console_auth_tokens.py b/nova/db/main/migrations/versions/2903cd72dc14_add_tls_port_to_console_auth_tokens.py
index 5d04018..e85a38e 100644
--- a/nova/db/main/migrations/versions/2903cd72dc14_add_tls_port_to_console_auth_tokens.py
+++ b/nova/db/main/migrations/versions/2903cd72dc14_add_tls_port_to_console_auth_tokens.py
@@ -29,6 +29,8 @@ depends_on = None


 def upgrade():
+    # NOTE(mikal): there is no shadow table for console_auth_tokens as there is
+    # no soft delete for these tokens.
     with op.batch_alter_table('console_auth_tokens', schema=None) as batch_op:
         batch_op.add_column(
             sa.Column('tls_port', sa.Integer(), nullable=True))
\ No newline at end of file
diff --git a/nova/objects/fields.py b/nova/objects/fields.py
index 332acf9..0747224 100644
--- a/nova/objects/fields.py
+++ b/nova/objects/fields.py
@@ -1209,6 +1209,26 @@ class XenAddress(AddressBase):
         return AddressBase.coerce(XenAddress, attr, value)


+sb16, es1370, pcspk, ac97 (Since 0.6.0), ich6 (Since 0.8.8), ich9 (Since 1.1.3),
+usb (Since 1.2.8), ich7 (Since 6.7.0, bhyve only) and virtio (Since 10.4.0 and QEMU 8.2.0).
+
+class AudioModel(BaseNovaEnum):
+
+    # NOTE(mikal): ich7 is not included because at the time of writing it is
+    # only supported by bhyve
+    SB16 = 'sb16'
+    ES1370 = 'es1370'
+    PCSPK = 'pcspk'
+    AC97 = 'ac97'
+    ICH6 = 'ich6'
+    ICH9 = 'ich9'
+    USB = 'usb'
+    VIRTIO = 'virtio'
+    NONE = 'none'
+
+    ALL = (SB16, ES1370, PCSPK, AC97, ICH6, ICH9, USB, VIRTIO, NONE)
+
+
 class USBAddressField(AutoTypedField):
     AUTO_TYPE = USBAddress()

@@ -1431,3 +1451,7 @@ class ListOfListsOfStringsField(AutoTypedField):

 class DictOfSetOfIntegersField(AutoTypedField):
     AUTO_TYPE = Dict(Set(fields.Integer()))
+
+
+class AudioModelField(BaseEnumField):
+    AUTO_TYPE = AudioModel()
\ No newline at end of file
diff --git a/nova/objects/image_meta.py b/nova/objects/image_meta.py
index ab486b3..1d8031d 100644
--- a/nova/objects/image_meta.py
+++ b/nova/objects/image_meta.py
@@ -196,14 +196,17 @@ class ImageMetaProps(base.NovaObject):
     # Version 1.36: Added 'hw_maxphysaddr_mode' and
     #                     'hw_maxphysaddr_bits' field
     # Version 1.37: Added 'hw_ephemeral_encryption_secret_uuid' field
+    # Version 1.38: Added 'hw_audio_model' field
     # NOTE(efried): When bumping this version, the version of
     # ImageMetaPropsPayload must also be bumped. See its docstring for details.
-    VERSION = '1.37'
+    VERSION = '1.38'

     def obj_make_compatible(self, primitive, target_version):
         super(ImageMetaProps, self).obj_make_compatible(primitive,
                                                         target_version)
         target_version = versionutils.convert_version_to_tuple(target_version)
+        if target_version < (1, 38):
+            primitive.pop('hw_audio_model', None)
         if target_version < (1, 37):
             primitive.pop('hw_ephemeral_encryption_secret_uuid', None)
         if target_version < (1, 36):
@@ -495,6 +498,9 @@ class ImageMetaProps(base.NovaObject):
         # Control bits for the physical memory address bit of Libvirt guests.
         'hw_maxphysaddr_bits': fields.IntegerField(),

+        # Name of the model of audio device to use, if any
+        'hw_audio_model': fields.AudioModelField(),
+
         # if true download using bittorrent
         'img_bittorrent': fields.FlexibleBooleanField(),

